### PR TITLE
Provide an ansible role for deployment and configuration

### PR DESCRIPTION
**Pull request by C0nsultant**
I've really grown fond of the project.
After installing it manually twice, it is now time to automate the process (docker is not an option in my case).
This is largely based on the official [bare-metal route](https://paperless-ng.readthedocs.io/en/latest/setup.html#bare-metal-route).

Caveats
- Currently only Debian 10 buster is supported
- Intended to be installed in _dirty_ enviroments (with other users/services)
	- dedicated user for paperless-ng
	- .venv
	- dropped permissions of the services
- Database deployment is considered out-of-scope and deferred to the user
- Provides basic upgrade support between releases
- `jbig2enc` is pulled from https://notesalexp.org/ (owned by @AlexanderP)

ToDo
- [x] Put `paperless.conf` in `/etc`
- [x] Fix `PAPERLESS_OCR_USER_ARG` construction
- [ ] Document installation
- [x] Test the role using https://github.com/ansible-community/molecule
- [ ] Provide molecule scenario to verify upgrading from last release works
- [ ] Publish the role to [ansible galaxy](https://galaxy.ansible.com/)
- [x] Harden systemd service files
- [ ] (Probably in a seperate PR) introduce more operating systems

### CODE REVIEW

**Comment by jonaswinkler** [comment]
Thank you!

I'll try to verify that this works once you're done.

File: ansible/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+paperlessng_data_dir: "{{ paperlessng_directory }}/data"
+paperlessng_media_root: "{{ paperlessng_directory }}/media"
+paperlessng_static_dir: "{{ paperlessng_directory }}/static"
+paperlessng_filename_format:
+paperlessng_virtualenv: "{{ paperlessng_directory }}/.venv"
+
+paperlessng_ocr_languages:
+  - eng
+paperlessng_time_zone: Europe/Berlin
+paperlessng_ocrmypdf_args: --optimize 1
```
Comment (jonaswinkler) [inline ansible/defaults/main.yml:14]: This is not how this works, sadly.

I'm using the API of ocrmypdf, and this requires arguments to be passed as kwargs to a python function. Therefore, this should be a JSON string with the arguments as a dictionary, see the documentation.

Comment (C0nsultant) [reply ansible/defaults/main.yml:14]: Thanks for the head-up! So the `# TODO JSON dict required in conf?` in the `configure paperless-ng` step is actually a problem... I'll add it to the list.
If you have any suggestions on how to solve this, please share.

File: ansible/tasks/main.yml [hunk 2]
```diff
@@ -209,17 +209,21 @@
...
 
 - name: collect static files
-  command: "{{ paperlessng_virtualenv }}/bin/python3 {{ paperlessng_directory }}/src/manage.py collectstatic --clear --no-input"
+  command: "{{ paperlessng_virtualenv }}/bin/python3 manage.py collectstatic --clear --no-input"
+  args:
+    chdir: "{{ paperlessng_directory }}/src"
 
 - name: create database schema
-  command: "{{ paperlessng_virtualenv }}/bin/python3 {{ paperlessng_directory }}/src/manage.py migrate"
+  command: "{{ paperlessng_virtualenv }}/bin/python3 manage.py migrate"
```
Comment (C0nsultant) [inline ansible/tasks/main.yml:217]: This command depends on the current working directory.
When in the default ansible temporary directory, `paperless.conf` wasn't read.
Not sure about the other `manage.py` calls.

```
SystemCheckError: System check identified some issues:

ERRORS:
?: PAPERLESS_CONSUMPTION_DIR is set but doesn't exist.
	HINT: Create a directory at /opt/paperless-ng/src/../consume
```

File: ansible/defaults/main.yml (hunk 1 above)
Comment (jonaswinkler) [reply ansible/defaults/main.yml:14]: Regarding ocrmypdf, I'll look into using their command line args parser and MAYBE I can get it to work. However, that's not part of their public api, so its really not a good idea.

I can see how it's cumbersome to have that specified as a json dict.

File: ansible/tasks/main.yml (hunk 2 above)
Comment (jonaswinkler) [reply ansible/tasks/main.yml:217]: Maybe move the config file to /etc/paperless.conf?

It's safer there anyway, since the paperlessng user won't be able to modify it.

File: ansible/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    creates: "{{ paperlessng_virtualenv }}"
+
+- name: install paperlessng requirements
+  pip:
+    requirements: "{{ paperlessng_directory }}/requirements.txt"
+    virtualenv: "{{ paperlessng_virtualenv }}"
+    extra_args: --upgrade
+
+- name: collect static files
+  command: "{{ paperlessng_virtualenv }}/bin/python3 manage.py collectstatic --clear --no-input"
```
Comment (C0nsultant) [inline ansible/tasks/main.yml:212]: Is there any reason why this should use `--clear` by deafult?
The [docs](https://docs.djangoproject.com/en/3.1/ref/contrib/staticfiles/#collectstatic) state that (without `--clear`) only files with differing timestamps are collected.

Comment (jonaswinkler) [reply ansible/tasks/main.yml]: Probably not. There won't be any static files on fresh installs in the first place. I'll also have that removed from the Dockerfiles. I've left that in there from paperless and did not really bother up until now.

**Comment by jonaswinkler** [comment]
Great, I'll fire up a debian 10 virtual machine and see how this goes.

Regarding user args, using the argument parser from ocrmypdf comes with some oddities and I'll have to check in with the dev over there. Might take a while. For now, this will remain a json string.

File: ansible/defaults/main.yml (hunk 1 above)
Comment (C0nsultant) [reply ansible/defaults/main.yml:14]: The json form does not really bother me TBH.
Configuring the whole OCRmyPDF backend by mapping a function call with [roughly 50](https://ocrmypdf.readthedocs.io/en/latest/api.html#ocrmypdf.ocr) arguments to a nicer structure will take more effort than I initially thought, though.
For now, I will just pass any json from the ansible variables to the `.conf`.

Off-topic: Do you think having this configureable in the frontend would be worth the time?

File: ansible/tasks/main.yml (hunk 2 above)
Comment (C0nsultant) [reply ansible/tasks/main.yml:248]: So there might actually be [three](https://github.com/jonaswinkler/paperless-ng/blob/ng-0.9.9/src/paperless/settings.py#L10-L15) possible locations.
`/etc/paperless.conf` is probably the best choice.

Comment (C0nsultant) [reply ansible/tasks/main.yml:248]: Two things that come to mind.

1. There should always be a valid `paperless.conf` at some fixed location. This currently is not the case (`/etc/paperless.conf` `/usr/local/etc/paperless.conf` aren't created if they don't exist, `../paperless.conf` only exists after templating it from `paperless.conf.example`).
2. If paperless would fall back to defaults because no `paperless.conf` can be found, the user should either be notified or the whole process should fail (my preference).

**Comment by jonaswinkler** [comment]
I'm getting a `ERROR! 'fail' is not a valid attribute for a Play` when trying to execute `ansible/tasks/main.yml` as a playbook with Ansible version 2.7.7.

Not sure if I'm supposed to do this, never used Ansible before.

**Comment by C0nsultant** [comment]
You appear to have run something like `ansible-playbook paperless-ng/ansible/tasks/main.yml`.
That won't work, unfortunately. What this PR provides is an ansible [**role**](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html).

**quick ansible 101**
Essentially, a role is a combination of ordered tasks/commands, variables and file templates.
Ansible then applies these roles to target machines they call [**hosts**](https://docs.ansible.com/ansible/latest/user_guide/intro_patterns.html). Hosts can also be grouped into - you guessed it - **groups**.
To manage large numbers of hosts and groups you structure them into a a hierarchical collection called [**inventory**](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html).

Now what you did there with `ansible-playbook` is the next step up.
[**playbooks**](https://docs.ansible.com/ansible/latest/user_guide/playbooks.html) combine inventories and roles to form larger deployment constructs with multiple setups across multiple hosts.
In the case of NG for example, you might have a playbook where your target host is your local machine, a remote server (or even both for a redundant setup). That playbook first sets up a PostgreSQL server using a role (e.g. https://github.com/geerlingguy/ansible-role-postgresql). After that has completed successfully, it proceeds to install/update paperless-ng using the role in the PR.

You could do [ad-hoc](https://docs.ansible.com/ansible/latest/user_guide/intro_adhoc.html) commands to test a single role, but I honstely haven't every bothered to do so.
playbooks document what you do way better than your shell history.

I would recommend writing a tiny playbook to execute the role on the target system of your choice.

**TL;DR**
> playbook.yml
```yaml
---
- name: run paperless-ng ansible role on current machine
  hosts: localhost
  gather_facts: true

  tasks:
    - name: "Include ansible"
      include_role:
        name: "ansible"  # this imports the role based on the folder name, something like `jonaswinkler.paperless-ng` once published
```

```sh
$ ansible-playbook playbook.yml
```

**Comment by C0nsultant** [comment]
Turns out the role wasn't compatible with `2.7.7` as the orchestrating ansible (I only tested with [2.9.16](https://koji.fedoraproject.org/koji/buildinfo?buildID=1660020) on Fedora 33 and [2.9.6] on the 20.04 Travis build).
Now it works on a bare Debian 10 installation with only ansible installed from the default sources.

File: ansible/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+paperlessng_static_dir: "{{ paperlessng_directory }}/static"
+paperlessng_filename_format:
+paperlessng_virtualenv: "{{ paperlessng_directory }}/.venv"
+
+paperlessng_ocr_languages:
+  - eng
+paperlessng_time_zone: Europe/Berlin
+# see https://ocrmypdf.readthedocs.io/en/latest/api.html#ocrmypdf.ocr
+paperlessng_ocrmypdf_args:
+  - "deskew": true
```
Comment (C0nsultant) [inline ansible/defaults/main.yml:16]: Setting `deskew` is a problem when the parser upgrades to `redo`. 
> --redo-ocr is not currently compatible with --deskew, --clean-final, and --remove-background

Comment (jonaswinkler) [reply ansible/defaults/main.yml:16]: I see. Well, this was done to cover some special PDF documents. I've got some documents over here that do have embedded and selectable text, but any of the python libraries that extract text from PDF documents fail to pick that up. After --redo-ocr, that works just fine.

So, what shall we do?

Comment (C0nsultant) [reply ansible/defaults/main.yml:16]: I also would like if active `deskew` could be the default without having to worry about breaking things.

But it is honestly not a problem paperless should have to worry about. The word **currently** in OCRmyPDF's error message clearly states that this was considered. Before we figure out a smart workaround, we should ask them about it.

Comment (C0nsultant) [reply ansible/defaults/main.yml]: Opened an issue at https://github.com/jbarlow83/OCRmyPDF/issues/708.

File: .github/workflows/ansible.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          python3 -m pip install --upgrade pip
+          python3 -m pip install molecule[ansible,docker]
+          ansible --version
+          docker --version
+          molecule --version
+          python --version
+      - name: Test with molecule
+        run: |
+          cd ansible
+          molecule test
```
Comment (C0nsultant) [inline .github/workflows/ansible.yml:32]: Testing the role in a PR with the latest release does not make sense.
What this needs is an option to pull not only a tagged release, but also a branch.

**Comment by jonaswinkler** [comment]
Alright, I just looked into this and getting this running on a fresh system is pretty simple. Adding documents works, which means that all of the important components are running properly.

Thank you for your explanations on ansible basics as well!

One thing you'll need to add for localization support:

```python manage.py compilemessages```

will build `django.mo` files from `django.po` files in the `src/locale/` folder. These are derived resources and therefore not in the repository. This should be done when you're also doing `collectstatic` and `migrate`.

How should we proceed? I can merge this if you want and you can push further changes with new PRs. Some documentation would be nice eventually.

**Comment by C0nsultant** [comment]
> How should we proceed? I can merge this if you want and you can push further changes with new PRs. 

Merging the current state should be the best way to go forward with this.
#304 was the next thing I would have touched, but I'll defer that to after #151.
So the next steps after this are
- a section in the documentation and
- a test to verify that upgrading between releases works.

One major problem I have run into: [`ansible-galaxy import`](https://galaxy.ansible.com/docs/contributing/importing.html) (the PyPI of ansible) pulls the contents for a role directly from git. A requirement for this to work is that the role sits at the root of a repository (instead of in a subfolder like in this PR).
There are a few options on how to solve this (if we even want to publish it to galaxy):
- put the role on its own branch (not my preference. having both in the same branch means they are versioned together, which seems like a good idea)
- put the role in a separate repository, maybe import it as a submodule
- keep the current structure, push the `ansible` directory to its own repository when publishing a new tag, and import to galaxy from that repository (personal preference since it does the best of both worlds)
